server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
             }
         }
     }
+
+    private class ConnectionHistory {
+        record NodeConnectionHistory(String ephemeralId, long disconnectTime, Exception disconnectCause) {}
+
+        /**
+         * Holds the DiscoveryNode nodeId to connection history record.
+         *
+         * Entries for each node are reserved during NodeConnectionsService.connectToNodes, by placing a (nodeId, dummy) entry
+         * for each node in the cluster. On node disconnect, this entry is updated with its NodeConnectionHistory. On node
+         * connect, this entry is reset to the dummy value. On NodeConnectionsService.disconnectFromNodesExcept, node entries
+         * are removed.
+         *
+         * Each node in the cluster always has a nodeHistory entry that is either the dummy value or a connection history record. This
+         * allows node disconnect callbacks to discard their entry if the disconnect occurred because of a change in cluster state.
+         */
+        private final NodeConnectionHistory dummy = new NodeConnectionHistory("", 0, null);
+        private final ConcurrentMap<String, NodeConnectionHistory> nodeHistory = ConcurrentCollections.newConcurrentMap();
+
+        ConnectionHistory() {
+            NodeConnectionsService.this.transportService.addConnectionListener(new TransportConnectionListener() {
+                @Override
+                public void onNodeConnected(DiscoveryNode node, Transport.Connection connection) {
+                    // log case where the remote node has same ephemeralId as its previous connection
+                    // (the network was disrupted, but not the remote process)
+                    NodeConnectionHistory nodeConnectionHistory = nodeHistory.get(node.getId());
+                    if (nodeConnectionHistory != null) {
+                        nodeHistory.replace(node.getId(), nodeConnectionHistory, dummy);
+                    }
+
+                    if (nodeConnectionHistory != null
+                        && nodeConnectionHistory != dummy
+                        && nodeConnectionHistory.ephemeralId.equals(node.getEphemeralId())) {
+                        if (nodeConnectionHistory.disconnectCause != null) {
+                            logger.warn(
+                                () -> format(
+                                    "reopened transport connection to node [%s] "
+                                        + "which disconnected exceptionally [%dms] ago but did not "
+                                        + "restart, so the disconnection is unexpected; "
+                                        + "if unexpected, see [{}] for troubleshooting guidance",
+                                    node.descriptionWithoutAttributes(),
+                                    nodeConnectionHistory.disconnectTime,
```

Comment:
This'll show the absolute disconnect time in milliseconds (i.e. since 1970) whereas I think we want to see the duration between the disconnect and the current time.